beacon-chain/types/block.go

```diff
@@ -128,3 +140,104 @@ func (b *Block) Attestations() []*pb.AttestationRecord {
+		log.Debugf("Can not process a genesis block")
+		return false
+	}
+
+	if !b.isSlotValid() {
+		log.Debugf("slot of block is too high: %d", b.SlotNumber())
+		return false
+	}
+
+	for index, attestation := range b.Attestations() {
+		if !b.isAttestationValid(index, aState, cState) {
+			log.Debugf("attestation invalid: %v", attestation)
+			return false
+		}
+	}
+
+	return true
+}
+
+// isAttestationValid validates an attestation in a block
+func (b *Block) isAttestationValid(attestationIndex int, aState *ActiveState, cState *CrystallizedState) bool {
+	// Validate attestation's slot number has is within range of incoming block number.
+	slotNumber := int(b.SlotNumber())
+	attestation := b.Attestations()[attestationIndex]
+	if int(attestation.Slot) > slotNumber {
+		log.Debugf("attestation slot number can't be higher than block slot number. Found: %d, Needed lower than: %d",
+			attestation.Slot,
+			slotNumber)
+		return false
+	}
+	if int(attestation.Slot) < slotNumber-params.CycleLength {
+		log.Debugf("attestation slot number can't be lower than block slot number by one CycleLength. Found: %v, Needed greater than: %v",
+			attestation.Slot,
+			slotNumber-params.CycleLength)
+		return false
+	}
+
+	if attestation.JustifiedSlot > cState.LastJustifiedSlot() {
+		log.Debugf("attestation's last justified slot has to match crystallied state's last justified slot. Found: %d. Want: %d",
+			attestation.JustifiedSlot,
+			cState.LastJustifiedSlot())
+		return false
+	}
+
+	// TODO: Validate last justified block hash matches in the crystallizedState.
+
+	// Get all the block hashes up to cycle length.
+	parentHashes := aState.getSignedParentHashes(b, attestation)
+	attesterIndices, err := cState.getAttesterIndices(attestation)
+	if err != nil {
+		log.Debugf("unable to get validator committee: %v", attesterIndices)
+		return false
+	}
+
+	// Verify attester bitfields matches crystallized state's prev computed bitfield.
+	if !casper.AreAttesterBitfieldsValid(attestation, attesterIndices) {
+		return false
+	}
+
+	// TODO: Generate validators aggregated pub key.
+
+	// Hash parentHashes + shardID + slotNumber + shardBlockHash into a message to use to
+	// to verify with aggregated public key and aggregated attestation signature.
+	msg := make([]byte, binary.MaxVarintLen64)
+	var signedHashesStr []byte
+	for _, parentHash := range parentHashes {
+		signedHashesStr = append(signedHashesStr, parentHash[:]...)
+		signedHashesStr = append(signedHashesStr, byte(' '))
+	}
+	binary.PutUvarint(msg, attestation.Slot%params.CycleLength)
+	msg = append(msg, signedHashesStr...)
+	binary.PutUvarint(msg, attestation.ShardId)
+	msg = append(msg, attestation.ShardBlockHash...)
+
+	msgHash := blake2b.Sum512(msg)
+
+	log.Debugf("Attestation message for shard: %v, slot %v, block hash %v is: %v",
+		attestation.ShardId, attestation.Slot, attestation.ShardBlockHash, msgHash)
+
+	// TODO: Verify msgHash against aggregated pub key and aggregated signature.
```

Comment:
Is there a github issue created for this todo? Please reference issue number here